File: pages/LoginPage.js
const BaseElement = require('../framework/BaseElement.js')
const logger = require('../logger.js')
const BaseForm = require('../framework/BaseForm.js')
const TextBox = require('../framework/TextBox.js')
const Button = require('../framework/Button.js')

class LoginPage extends BaseForm {
  constructor() {
    super('.login_wrapper', 'Login Form')
    this.usernameField = new TextBox('//input[@data-test="username"]', 'Username Field')
    this.passwordField = new TextBox('//input[@data-test="password"]', 'Password Field')
    this.loginButton = new Button('//input[@data-test="login-button"]', 'Login Button')
    this.swagLabsHeader = new BaseElement('.login_logo', 'Swag Labs Header')
    this.credentialsBlock = new BaseElement('//div[@data-test="login-credentials-container"]', 'Credentials Block')
    this.errorMessage = new BaseElement('//h3[@data-test="error"]', 'Error Message')
    this.errorMessageExitButton = new Button('//button[@data-test="error-button"]/*[local-name()="svg"]', 'Error Message Exit Button')
    this.usernameErrorIcon = new Button(
      '//div[@class="form_group"]/input[@data-test="username"]/following-sibling::*[local-name()="svg"]',
      'Username Error Icon',
    )
    this.passwordErrorIcon = new Button(
      '//div[@class="form_group"]/input[@data-test="password"]/following-sibling::*[local-name()="svg"]',
      'Password Error Icon',
    )
    this.passwordsBlock = new TextBox('//div[@data-test="login-password"]', 'Credentials block passwords')
    this.usernamesBlock = new TextBox('//div[@data-test="login-credentials"]', 'Credentials block usernames')
  }

  async login(username, password) {
    await this.usernameField.sendText(username)
    await this.passwordField.sendText(password)
    await this.loginButton.click()
  }
  async enterPassword(password){
    await this.passwordField.sendText(password)
  }
  async enterUsername(username){
    await this.usernameField.sendText(username)
  }
  async submit(){
    await this.loginButton.click()
  }

  async isErrorMessageDisplayed() {
    return this.errorMessage.isDisplayed()
  }

  async isErrorIconUsernameDisplayed() {
    return this.usernameErrorIcon.isDisplayed()
  }

  async isErrorIconPasswordDisplayed() {
    return this.passwordErrorIcon.isDisplayed()
  }
  async getErrorMessageText(){
    return this.errorMessage.getText()
  }
  async clickErrorMessageExitButton(){
    await this.errorMessageExitButton.click()
  }
  

  async closeErrorMessage() {
    await this.errorMessageExitButton.click()
  }
  async getUsernamePlaceholder(){
    return this.usernameField.getAttribute('placeholder')
  }
  async getPasswordPlaceholder(){
    return this.passwordField.getAttribute('placeholder')
  }
  async getLoginButtonValue(){
    return this.loginButton.getValue()
  }
  async getLoginButtonColor(){
    return this.loginButton.getCSSValue('background-color')
  }
  async getUsernameBlockText(){
    return this.usernamesBlock.getText()
  }
  async getPasswordBlockText(){
    return this.passwordsBlock.getText()
  }
  async isHeaderDisplayed(){
    return this.swagLabsHeader.isDisplayed()
  }
  async getHeaderText(){
    return this.swagLabsHeader.getText()
  }
  async isUsernameFieldDisplayed(){
    return this.usernameField.isDisplayed()
  }
  async isPasswordFieldDisplayed(){
    return this.passwordField.isDisplayed()
  }
  async isLoginButtonDisplayed(){
    return this.loginButton.isDisplayed()
  }
  async isCredentialsBlockDisplayed(){
    return this.credentialsBlock.isDisplayed()
  }
  async isLoginButtonEnabled(){
    return this.loginButton.isEnabled()
  }
}

module.exports = LoginPage

File: src/test/java/pages/login/LoginPage.java
package pages.login;

import framework.utils.BaseForm;
import framework.elements.Label;
import org.openqa.selenium.By;

import java.util.Arrays;
import java.util.List;

public class LoginPage extends BaseForm {
    private final Label header;
    private final Label usernamesCredentialsBlock;
    private final Label passwordsCredentialsBlock;

    public LoginPage(){
        super(By.cssSelector("[data-test='login-container']"), "Login Page Container");
        this.header = new Label(By.cssSelector(".login_logo"), "Header Label");
        this.usernamesCredentialsBlock = new Label(By.xpath("//div[@data-test='login-credentials']"), "Usernames Credentials Block");
        this.passwordsCredentialsBlock = new Label(By.xpath("//div[@data-test='login-password']"), "Password Credentials Block");
    }

    //Header action methods
    public boolean IsHeaderDisplayed(){
        return header.IsVisible();
    }

    public String GetHeaderText(){
        return header.GetText();
    }

    //Credentials block methods
    public boolean IsCredentialsBlockDisplayed(){
        return usernamesCredentialsBlock.IsVisible() && passwordsCredentialsBlock.IsVisible();
    }

    public List<String> GetUsernamesList(){
        var namesArray = (usernamesCredentialsBlock.GetText()).split("\n");
        var namesList = Arrays.asList(namesArray);
        return namesList;
    }

    public List<String> GetPasswordsList(){
        var passwordsArray = (passwordsCredentialsBlock.GetText()).split("\n");
        var passwordsList = Arrays.asList(passwordsArray);
        return passwordsList;
    }
}
